include/drivers/gpio.h

```diff
@@ -568,14 +589,37 @@ static inline int z_impl_gpio_pin_interrupt_configure(struct device *port,
 {
 	const struct gpio_driver_api *api =
 		(const struct gpio_driver_api *)port->driver_api;
+	struct gpio_driver_data *const data =
+			(struct gpio_driver_data *const)port->driver_data;
 
 	__ASSERT(pin < GPIO_MAX_PINS_PER_PORT, "Invalid pin number");
 
 	__ASSERT(((flags & GPIO_INT_ENABLE) == 0) ||
 		 ((flags & (GPIO_INT_LOW_0 | GPIO_INT_HIGH_1)) != 0),
 		 "At least one of GPIO_INT_LOW_0, GPIO_INT_HIGH_1 has to be "
 		 "enabled.");
-	return api->pin_interrupt_configure(port, pin, flags);
+
+	flags = flags & GPIO_INT_ALL_MASK;
+	if ((flags & GPIO_INT_LEVELS_LOGICAL) && (data->invert & BIT(pin))) {
+		/* Invert signal bits */
+		flags = ~(flags & (GPIO_INT_LOW_0 | GPIO_INT_HIGH_1)) |
+			(flags & ~(GPIO_INT_LOW_0 | GPIO_INT_HIGH_1));
+	}
+
+	/* Clear GPIO_INT_LEVELS_LOGICAL */
+	flags &= ~GPIO_INT_LEVELS_LOGICAL;
+
+	if ((flags == GPIO_INT_MODE_DISABLED) ||
+	    (flags == GPIO_INT_MODE_LEVEL_LOW) ||
+	    (flags == GPIO_INT_MODE_LEVEL_HIGH) ||
+	    (flags == GPIO_INT_MODE_EDGE_FALLING) ||
+	    (flags == GPIO_INT_MODE_EDGE_RISING) ||
+	    (flags == GPIO_INT_MODE_EDGE_BOTH)) {
+		return api->pin_interrupt_configure(port, pin,
+				(enum gpio_int_trigger_mode)flags);
+	}
+
+	return -EINVAL;
 }
```

Comment:
I didn't analyse this code, however it doesn't pass the `tests/drivers/gpio/gpio_api_1pin` testcase. Instead the function can be written as
```
static inline int z_impl_gpio_pin_interrupt_configure(struct device *port,
		unsigned int pin, unsigned int flags)
{
	const struct gpio_driver_api *api =
		(const struct gpio_driver_api *)port->driver_api;
	struct gpio_driver_data *const data =
		(struct gpio_driver_data *const)port->driver_data;

	__ASSERT(pin < GPIO_MAX_PINS_PER_PORT, "Invalid pin number");

	__ASSERT(((flags & GPIO_INT_ENABLE) == 0) ||
		 ((flags & (GPIO_INT_LOW_0 | GPIO_INT_HIGH_1)) != 0),
		 "At least one of GPIO_INT_LOW_0, GPIO_INT_HIGH_1 has to be "
		 "enabled.");

	if (((flags & GPIO_INT_LEVELS_LOGICAL) != 0) &&
	    ((data->invert & BIT(pin)) != 0)) {
		/* Invert signal bits */
		flags ^= (GPIO_INT_LOW_0 | GPIO_INT_HIGH_1);
	}

	flags &= (GPIO_INT_LOW_0 | GPIO_INT_HIGH_1 | GPIO_INT_EDGE |
		  GPIO_INT_ENABLE);

	return api->pin_interrupt_configure(port, pin,
			(enum gpio_int_trigger_mode)flags);
}
```